test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/SemanticAst/ReferenceSegmentTests.cs

```diff
@@ -0,0 +1,57 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ReferenceSegmentTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using FluentAssertions;
+using Microsoft.OData.UriParser;
+using Xunit;
+
+namespace Microsoft.OData.Tests.UriParser.SemanticAst
+{
+    public class ReferenceSegmentTests
+    {
+        #region Test Cases
+        [Fact]
+        public void ReferenceSegmentConstructWithNullNavigationSourceShouldThrowException()
+        {
+            Action create = () => new ReferenceSegment(null);
+            create.ShouldThrow<ArgumentNullException>().WithMessage("Value cannot be null.\r\nParameter name: navigationSource");
+        }
+
+        [Fact]
+        public void ReferenceSegmentWithNavigationSourceConstructsSuccessfully()
+        {
+            ReferenceSegment referenceSegment = new ReferenceSegment(HardCodedTestModel.GetPet1Set());
+
+            referenceSegment.Identifier.Should().Be(UriQueryConstants.RefSegment);
+            referenceSegment.SingleResult.Should().BeFalse();
+            referenceSegment.TargetEdmNavigationSource.Should().Be(HardCodedTestModel.GetPet1Set());
+            referenceSegment.TargetKind.Should().Be(RequestTargetKind.Resource);
+            referenceSegment.EdmType.Should().BeNull();

```

Comment:
>referenceSegment [](start = 12, length = 16)

add validation that (in this case) referenceSegment.SingleResult.Should().BeTrue().  Add another test case with a single-valued nav source where referenceSegment.SingleResult.Should().BeFalse().

Reply:
Second one would be something like:
            ReferenceSegment referenceSegment = new ReferenceSegment(HardCodedTestModel.GetBossSingleton());
            referenceSegment.Identifier.Should().Be(UriQueryConstants.RefSegment);
            referenceSegment.SingleResult.Should().BeTrue();
            referenceSegment.TargetEdmNavigationSource.Should().Be(HardCodedTestModel.GetBossSingleton());
            referenceSegment.TargetKind.Should().Be(RequestTargetKind.Resource);

---
In reply to: [224945402](https://github.com/OData/odata.net/pull/1110#discussion_r224945402) [](ancestors = 224945402)

Reply:
Ah okay, that's added now.